README.md
# EMHMM Analysis

## `./bdd/`

Data and Results folder for EMHMM runs

## `./helpers/`

Helper functions. Remember to add this folder to your path.

## `./scripts/`

1. `fixations_preprocessing.m` adapted from Alice's `Convert_Fixation_2_Cocluster_Format.m` with snippets to map data in different blocks to the single corresponding participant.

2. `cocluster_categorization.m` the main co-clustering script adapted from `demo_cocluster_Categorization_final.m` from emhmm-toolbox and Jennifer's script. Perform 1-group/2-group co-clustering using different cases.

3. `remove_outliers.m` adapted from `vhem_plot_fixations.m` in the emhmm-toolbox that can be called within a 1-group co-clustering run using `cocluster_categorization.m`

4. `replace_group_hmms.m`. This script should be run after obtaining the cogroup hmms for specific edge case stimuli. Set the file name and index of the edge case among all stimuli. It will replace the corresponding parameters in the cogroup hmms of all cases.

### `./scripts/Jennier/`

Original co-clustering script used by Jennifer. Semantic analysis example from Jennifer.

### `./scripts/Alice`

Scripts shared by Alice that are not yet used.

# XAI

1. Script of generate human attention maps for similarity analysis: `H:\Projects\HKU_XAI_Project\XAI_Similarity_1\Main_gen_human_saliency_map_2.m`

# Paths

```
% Human
   
%Categorization
% OPT.fixationfile = 'bdd/fixation/hum_id_fix_12_17_2022_17_58.xlsx';
% OPT.imgdir       = 'bdd/images/orib_hum_id_task_resized/';
% OPT.imgsize      = [1024; 576];
% OPT.imginfo      = 'bdd/image_info/hum_id_stimuli_info.xlsx';
% OPT.outdir       = 'bdd/results/identification/ORIB-data-vb-results-lab-hum-local/';

%Explanation
% OPT.fixationfile = 'bdd/fixation/hum_exp_fix_10_04_2023_02_54.xlsx';
% OPT.imgdir       = 'bdd/images/orib_hum_id_task_resized/';
% OPT.imgsize      = [1024; 576];
% OPT.imginfo      = 'bdd/image_info/hum_id_stimuli_info.xlsx';
% OPT.outdir       = 'bdd/results/explanation/human/';

% Explanation whole screen
% OPT.fixationfile = 'bdd/fixation/hum_exp_screen_10_07_2023_23_49.xlsx';
% OPT.imgdir       = 'bdd/images/human_with_screen/';
% OPT.imgsize      = [1024; 768];
% OPT.imginfo      = 'bdd/image_info/hum_exp_screen_stimuli_info.xlsx';
% OPT.outdir       = 'bdd/results/explanation/231015_human_whole_screen_1-group/';
% OPT.HEM_K         = 1;    % fit one group to identify outliers

% Vehicle

%Categorization
% OPT.fixationfile = 'bdd/fixation/veh_id_fix_12_07_2022_11_00.xlsx';
% OPT.imgdir       = 'bdd/images/orib_veh_id_task_resized/';
% OPT.imgsize      = [1024; 576];
% OPT.imginfo      = 'bdd/image_info/veh_id_stimuli_info.xlsx';
% OPT.outdir       = 'bdd/results/identification/ORIB-data-vb-results-lab-veh-300-rerun/';

%Explanation
% OPT.fixationfile = 'bdd/fixation/veh_exp_screen_10_07_2023_23_49.xlsx';
% OPT.imgdir       = 'bdd/images/orib_veh_id_task_resized/';
% OPT.imgsize      = [1024; 576];
% OPT.imginfo      = 'bdd/image_info/veh_id_stimuli_info.xlsx';
% OPT.outdir       = 'bdd/results/explanation/vehicle/';

% Explanation whole screen
OPT.fixationfile = 'bdd/fixation/veh_exp_screen_10_07_2023_23_49.xlsx';
OPT.imgdir       = 'bdd/images/vehicle_with_screen/';
OPT.imgsize      = [1024; 768];
OPT.imginfo      = 'bdd/image_info/veh_exp_screen_stimuli_info.xlsx';
OPT.outdir       = 'bdd/results/explanation/231015_vehicle_whole_screen_1-group/';
OPT.HEM_K         = 1;    % fit one group to identify outliers


%Categorization and Explanation
% OPT.fixationfile = 'Input-Data/all-cat-exp-data';
% OPT.imgdir       = 'HUAWEI-categorization-data/';
% OPT.imgsize      = [1024; 576];
% OPT.imginfo      = 'HUAWEI-categorization-data/stimuli-info.xlsx';
% OPT.outdir       = 'Categorization-Explanation-results/';
```

emhmm-toolbox/CHANGELOG.txt
========================================================================

  Eye-Movement analysis with Hidden Markov Models (HMMs)
  emhmm-toolbox

  Antoni B. Chan, City University of Hong Kong
  Janet H. Hsiao, University of Hong Kong
  Tim Chuk, University of Hong Kong
  Lan Hui, City University of Hong Kong

  Copyright (c) 2021, City University of Hong Kong & University of Hong Kong

========================================================================

--- CHANGE LOG ---
v0.80 2021/09/10
- added co-clustering (handles stimuli with different configurations)
- added VB co-clustering (co-clustering w/ automatic selection of numbers of groups and ROIs)
- added co-clustering and VB co-clustering demo
- added VB clustering (automatically select numbers of groups for clustering)
- added VB clustering demos
- new support functions: read in data for co-clustering (read_xls_fixations2, preprocess_fixations)
- new support functions: save figure as image (savefigs), print cell table

v0.78 2021/07/20
- new functions for computing conditional and marginal entropies of an HMM
- new demo for computing entropy (demo_faces_entropy)
- new function to permute VHEM clusters (vhem_permute_clusters)
- new function to compute symmetric KL divergence (stats_ttest_skl)
- vbhmm_learn now supports specifying a fixed set of ROIs (fixed_rois option)
- stats_ttest computes Cohen's d
- read_xls_fixations now supports ignoring fixations outside of rectangular or elliptical region(s).
- vbhmm_plot_compact, vhem_plot, and vhem_plot_fixations - added option to plot transition matrix on the bottom.
- added FAQ in documentation.
- various bug fixes
- Note: co-clustering code will be released in the next version.

v0.77 2020/03/12
- beta code for co-clustering (cocluster directory)
- various bug fixes

v0.76 2019/06/15
- for CogSci2019 tutorial: new demo and tutorial slides
- various bug fixes

v0.75 2019/02/22
- use parallel for loop (parfor) to improve speed (if Parallel Computing Toolbox is available).
- add vbopt.seed and hemopt.seed to seed the random number generator for reproducible results.
- NOTE: an error will occur if you don't set the seed explicitly in vbopt/hemopt.
- updated demo files to use vbopt.seed and hemopt.seed.
- NOTE: you will need to update your experiment scripts to use vbopt.seed and hemopt.seed.
- improved visualizations: text in figures now scale with the figure size.

v0.74 2019/01/12 (internal)
- can pass images directly to plotting functions.
- grayscale option for figures
- new functions: vbhmm_entropy, vbhmm_random_sample
- handle empty HMMs in VHEM
- fixed consistency with older versions of MATLAB.
- improve numerical stability.

v0.73 2018/06/13
- vbhmm_standardize: add option to sort ROIs by steady-state probabilities of ROIs
- vhem_cluster: add option to remove empty ROIs in the input HMMs (set by default)
- new functions: vbhmm_prob_steadystate
- bug fixes: VHEM with one ROI or tau=1 (thanks, Sara!)
- internal cleanup

v0.72 2018/05/19
- added holistic/analytic representative models from PBR paper
- support for MEX files in Linux
- updated support for fixed groups is vbhmm_learn
- new hyp optimization option: added 'W0log' and min/max hyp values
- added support functions: vbhmm_remove_empty
- plotting: added color labels for states
- bug fixes: check for valid v0 hyperparameter
- various improvements to numerical stability

v0.71 2018/02/01
- small bug fixes (Thanks Joe and Louis!)

v0.70 2017/08/26
- added automatic estimation of HMMs hyperparameters (vbhmm_learn, vbhmm_learn_batch)
- added automatic selection of initialization for VHEM clustering.
- added full covariance matrices for VHEM clustering (now default)
- removed dependency on Image Processing toolbox.
- improved speed using MEX files and other MATLAB tricks.

v0.64 2016/06/07
- (internal release)
- bug fixes.
- code cleanup for VHEM.

v0.63 2017/05/25
- added check for updates.
- added stats_meanll - calculate mean LL for each subject.
- added stats_stateseq - find the most likely ROI sequences.
- updated face_demo to show the above two new analyses.
- added conversion between different face images, and associated demo.
- bug fixes.

v0.62 2017/05/24
- added support for more colors when plotting.
- better plotting for large number of ROIs.

v0.61 2017/05/17
- BUG FIX in vbhmm_ll - length-1 sequences were accidently normalized as length-2 sequences.
  This may affect some log-likelihood correlation tests. T-tests should not be affected much
  since the difference between data LL of two HMMs is used.
- remove vbhmm_ll_OLD
- add support for 'gmm' initialization in VHEM-HMM

v0.6 2017/02/13
- minor bug fixes in vbhmm_fb (doesn't affect results)
- code cleanup for vbhmm_em

v0.5 2016/01/21
- updated to handle fixation duration with xy; bug fixes; automatically select some hyperparameters.
- updated for Antoine: options for gmdistribution.fit (random_gmm_opt); 

v0.4 2016/01/13
- added xls reader and demo; use random initialization; added documentation (docs)

v0.3 2016/12/09
- added more demos, added component splitting for initialization (random init is still better).

v0.2 2016/11/21
- added HEM clustering, ttest for HMMs, rearranged source files, added demo code

v0.1 2016/08
- initialize VB with ROIs, and use fixed ROIs

v0.0 2016/06/17
- distribution 20160617

v0.0 2016/06
- added VB-HMM for learning individual's HMMs


emhmm-toolbox/src/hmm/vbhmm_fb.m
function [fbstats] = vbhmm_fb(data, hmm_varpar, opt)
% vbhmm_fb - run forward-backward algorithm
%
% this is an internal function called by vbhmm_em
%
% [fbstats] = vbhmm_fb(data, hmm_varpar, opt)
%
% opt.savexi = 1: also save xi for each t (default=0)
% opt.usegroups = 1: use groups mapping (default=0)
% opt.useMEX = 1: use MEX implementation if available (default=1)
%
% ---
% Eye-Movement analysis with HMMs (emhmm-toolbox)
% Copyright (c) 2017-01-13
% Antoni B. Chan, Janet H. Hsiao, Tim Chuk
% City University of Hong Kong, University of Hong Kong

% 2017-02-13: ABC: code cleanup (rename some variables). Speed optimization.
% 2017-06-09: ABC: code cleanup (rename some variables).
% 2017-06-09: ABC: added option to calculate derivatives
% 2017-06-14: ABC: code cleanup: epsilon & logATilde changed to row format
% 2017-07-31: ABC: removed calculate derivative options, since it's not used anymore.
% 2017-08-01: ABC: speed-up using MEX implementation
% 2017-08-03: ABC: clean-up code
% 2017-08-25: ABC: added scaling for numerical precision problems
% 2018-03-10: ABC: support groups

% hmm parameters: v, W, epsilon, alpha, m, beta
v       = hmm_varpar.v;
W       = hmm_varpar.W;
epsilon = hmm_varpar.epsilon;
alpha   = hmm_varpar.alpha;
m       = hmm_varpar.m;
beta    = hmm_varpar.beta;

% options: group_map, usegroups
usegroups = opt.usegroups;
if usegroups
  group_inds = opt.group_inds;
  group_map = opt.group_map;
  numgroups = opt.numgroups;
end

% other options: save xi
if isfield(opt, 'savexi')
  savexi = opt.savexi;
else
  savexi = 0;
end
if ~isfield(opt, 'useMEX')
  opt.useMEX = 1;
end

% some constants: K, dim, N, maxT, lengthT, const, const_denominator
K   = size(m, 2);
dim = size(m, 1);
N   = length(data);
maxT    = max(cellfun('prodofsize', data)/dim);
lengthT = maxT;
const = dim*log(2);
const_denominator = (dim*log(2*pi))/2;

% pre-calculate constants
logLambdaTilde = zeros(1,K);
for k = 1:K
  t1 = psi(0, 0.5*repmat(v(k)+1,dim,1) - 0.5*[1:dim]');
  logLambdaTilde(k) = sum(t1) + const  + log(det(W(:,:,k)));
end

if ~usegroups
  psiEpsilonHat = zeros(1,K);
  logATilde     = zeros(K,K);
  for k = 1:K
    %Bishop (10.66)
    psiEpsilonHat(k) = psi(0,sum(epsilon(k,:)));
    logATilde(k,:) = psi(0,epsilon(k,:)) - psiEpsilonHat(k);  % A(i,j) = p(j->i) [column]
  end
  psiAlphaHat = psi(0,sum(alpha));
  logPiTilde = psi(0,alpha) - psiAlphaHat;
  
else
  psiEpsilonHat = {};   logATilde = {};
  psiAlphaHat = {};     logPiTilde = {};
  for g = 1:numgroups
    for k = 1:K
      %Bishop (10.66)
      psiEpsilonHat{g}(k) = psi(0,sum(epsilon{g}(k,:)));
      logATilde{g}(k,:) = psi(0,epsilon{g}(k,:)) - psiEpsilonHat{g}(k);  % A(i,j) = p(j->i) [column]
    end
    psiAlphaHat{g} = psi(0,sum(alpha{g}));
    logPiTilde{g} = psi(0,alpha{g}) - psiAlphaHat{g};
  end
end


%% check if MEX can be used
% MEX does not support savexi
canuseMEX = 0;
if (opt.useMEX)
  if (exist('vbhmm_fb_mex')==3)
    canuseMEX = 1;
    if (savexi)
      canuseMEX = 0;
      warning('vbhmm_fb: savexi not supported in MEX implementation');
    end
    %if (usegroups)
    %  canuseMEX = 0;
    %  warning('vbhmm_fb: usegroups not supported in MEX implementation');
    %end
  else
    canuseMEX = 0;
    warning('vbhmm_fb: MEX implementation not found');
  end
end

if (opt.useMEX) && (canuseMEX)
  %% use MEX implementation
  
  if ~usegroups  
    % setup some variables first
    t_pz1     = exp(logPiTilde'); % priors p(z1)
    t_tpztzt1 = exp(logATilde); % transitions p(zt|z(t-1)) [row format]
    
    % Inputs
    %   data = {Nx1} - data{i} = [dim x T]
    %   K, N, dim, maxT = scalars [1x1]
    %   m = [dim x K]
    %   W = [dim x dim x K]
    %   v = [K x 1]
    %   beta = [K x 1]
    %   logLambdaTilde = [1 x K]
    %   const_denominator = scalar [1 x 1]
    %   t_pz1 = [1 x K]
    %   t_tpz1zt1 = [K x K]
    %
    % Outputs:
    %   logrho_Saved = [K x N x maxT]
    %   gamma_all    = [K x N x maxT]
    %   xi_sum       = [K x K x N]
    %   phi_norm     = [1 x N]
    %   xi_Saved (unsupported, since currently unused elsewhere)
    
    % call MEX function
    [logrho_Saved, gamma_all, xi_sum, phi_norm] = ...
      vbhmm_fb_mex(data, K, N, dim, maxT, m, W, v, beta, logLambdaTilde, const_denominator, t_pz1, t_tpztzt1);
    
  else
    % use groups
    
    % storage
    logrho_Saved = zeros(K, N, lengthT);
    phi_norm = zeros(1,N);
    gamma_all = zeros(K,N,maxT);
    xi_sum = zeros(K,K,N);
  
    for g=1:numgroups
      % run forward-backward for data in each group
      g_inds = group_inds{g};
      
      % get params for this group
      t_pz1      = exp(logPiTilde{g}');  % priors p(z1)
      t_tpztzt1  = exp(logATilde{g});    % transitions p(zt|z(t-1)) [row format]
      
      % select data for this group
      g_data = data(g_inds);
      g_N = length(g_data);
    
      % run MEX file
      [g_logrho_Saved, g_gamma_all, g_xi_sum, g_phi_norm] = ...
        vbhmm_fb_mex(g_data, K, g_N, dim, maxT, m, W, v, beta, logLambdaTilde, const_denominator, t_pz1, t_tpztzt1);
    
      % store results
      logrho_Saved(:,g_inds,:) = g_logrho_Saved;
      gamma_all(:,g_inds,:)    = g_gamma_all;
      phi_norm(g_inds)         = g_phi_norm;
      xi_sum(:,:,g_inds)       = g_xi_sum;
    end
    
    % check w/ non-MEX version    
    %tmpopt = opt;
    %tmpopt.useMEX = 0;
    %tmpfbstats = vbhmm_fb(data, hmm_varpar, tmpopt);
    %
    %e1 = sum(abs(tmpfbstats.logrho_Saved(:) - logrho_Saved(:)));
    %e2 = sum(abs(tmpfbstats.gamma_all(:) - gamma_all(:)));
    %e3 = sum(abs(tmpfbstats.phi_norm(:) - phi_norm(:)));
    %e4 = sum(abs(tmpfbstats.xi_sum(:) - xi_sum(:)));
    %
    %tmp = [e1, e2, e3, e4];
    %if any(tmp > 1e-9)
    %  keyboard
    %end    
  end
    
else
  if (opt.useMEX)
    % message if we wanted to use MEX but couldn't
    warning('vbhmm_fb: using MATLAB implementation. This will be slower');
  end
  
  %% MATLAB implementation
  
  % initialize storage
  logrho_Saved = zeros(K, N, lengthT);
  phi_norm = zeros(1,N);
  gamma_all = zeros(K,N,maxT);
  xi_sum = zeros(K,K,N);
  if savexi
    xi_Saved = cell(1,N);
    for n=1:N
      xi_Saved{n} = zeros(K,K,size(data{n},1));
    end
  end
  
  % precompute pz1 and tpztzt1
  if ~usegroups
    t_pz1     = exp(logPiTilde'); % priors p(z1)
    t_tpztzt1 = exp(logATilde);   % transitions p(zt|z(t-1)) [row format]
  else
    for g=1:numgroups   
      gt_pz1{g}         = exp(logPiTilde{g}');  % priors p(z1)
      gt_tpztzt1{g}     = exp(logATilde{g});    % transitions p(zt|z(t-1)) [row format]
    end    
  end
  
  %% run FB on each sequence
  for n = 1:N
    tdata = data{n}; tdata = tdata';
    tT = size(tdata,2);
    %delta = []; logrho = [];
    
    delta = zeros(K,tT);
    
    for k = 1:K
      %OLD slow code Bishop (10.64)
      %for t = 1:tT
      %  diff = tdata(:,t) - m(:,k);
      %  delta(k,t) = dim/beta(k) + v(k)*diff'*W(:,:,k)*diff;
      %end
      %delta_old = delta;
      
      % ABC: 2016-04-21 - fast code
      diff = bsxfun(@minus, tdata, m(:,k));
      mterm = sum((W(:,:,k)*diff).*diff,1);
      delta(k,:) = dim/beta(k) + v(k) * mterm;
      
    end
    % OLD slow code - Bishop (10.46)
    %for k = 1:K
    %  for t = 1:tT
    %    logrho(k,t) = 0.5*logLambdaTilde(k) - 0.5*delta(k,t) - const_denominator;
    %  end
    %end
    %logrho_old = logrho;
    
    % ABC: 2016-04-21 - fast code
    logrho = bsxfun(@minus, 0.5*logLambdaTilde(:), 0.5*delta) - const_denominator;
    
    logrho_Saved(:,n,1:tT) = logrho;
    
    %% forward_backward %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    %gamma = zeros(K,tT);
    %sumxi = zeros(K,K);  % [row]
    
    fb_logrho = logrho';
    %if ~usegroups
    %  fb_logPiTilde = logPiTilde';
    %  fb_logATilde = logATilde;
    %else
    %  fb_logPiTilde = logPiTilde{group_map(n)}';
    %  fb_logATilde = logATilde{group_map(n)};
    %end
    %t_alpha = [];t_beta = []; t_c = [];
    t_alpha = zeros(tT, K); % alpha hat
    t_Delta = zeros(tT, K); % Delta (unnormalized version)
    t_beta  = zeros(tT, K);
    t_Eta   = zeros(tT, K); % Eta (unnormalized version)
    t_c     = zeros(1,tT);
    
    %t_pz1 = exp(fb_logPiTilde);    % priors p(z1)
    %t_tpztzt1 = exp(fb_logATilde); % transitions p(zt|z(t-1)) [row format]
    if usegroups
      t_pz1 = gt_pz1{group_map(n)};
      t_tpztzt1 = gt_tpztzt1{group_map(n)};
    end
    
    %t_pxtzt = exp(fb_logrho);      % emissions p(xt|zt)
    
    % emission p(xt|zt)/max(p(xt|zt)) (more stable)
    max_fb_logrho = max(fb_logrho, [], 2);
    t_pxtzt = exp(bsxfun(@minus, fb_logrho, max_fb_logrho));
    
    t_x = tdata';
    t_T = size(t_x,1);
    
    t_gamma = zeros(K,t_T);
    t_sumxi = zeros(K,K);
      
    if t_T >= 1
      %forward
      %t_gamma = [];
      
      %before normalizing
      t_Delta(1,:) = t_pz1.*t_pxtzt(1,:);
      
      % 2016-04-29 ABC: rescale for numerical stability (otherwise values get too small)
      t_c(1) = sum(t_Delta(1,:));
      
      % normalize
      t_alpha(1,:) = t_Delta(1,:) / t_c(1);
      
      if t_T > 1
        for i=2:t_T
          % before normalizing
          t_Delta(i,:) = t_alpha(i-1,:)*t_tpztzt1.*t_pxtzt(i,:);
          
          % 2016-04-29 ABC: rescale for numerical stability
          t_c(i) = sum(t_Delta(i,:));
          
          % normalize
          t_alpha(i,:) = t_Delta(i,:) / t_c(i);
        end
      end
      
      %% backward %%%%%%%%%%%%%%%%%%%%%%
      % BUG FIX 2017-02-09: ones(1,K)./K;
      % doesn't affect HMM results, since we had normalized anyways.
      t_beta(t_T,:) = ones(1,K);
      
      t_gamma(:,t_T) = (t_alpha(t_T,:).*t_beta(t_T,:))';
      
      if t_T > 1
        for i=(t_T-1):-1:1
          % before normalizing
          bpi = (t_beta(i+1,:).*t_pxtzt(i+1,:));
          t_Eta(i,:) = bpi*t_tpztzt1';
          
          % 2016-04-29 ABC: rescale
          t_beta(i,:) = t_Eta(i,:)/t_c(i+1);
          
          % responsibility
          t_gamma(:,i) = (t_alpha(i,:).*t_beta(i,:))';
          
          %t_sumxi = t_sumxi + (t_logATilde.*(t_alpha(i,:)'*bpi));
          
          % 2016-04-29 ABC: rescale xi
          tmp_xi = (t_tpztzt1.*(t_alpha(i,:)'*bpi)) / t_c(i+1);
          
          % 2016-04-29 ABC BUG FIX: normalize xi matrix to sum to 1
          % (it's a joint probability matrix)
          % 2017-02-09 - not necessary anymore after fixing beta initialization bug.
          %tmp_xi = tmp_xi / sum(tmp_xi(:));
          
          % accumulate
          t_sumxi = t_sumxi + tmp_xi;
          
          if savexi
            xi_Saved{n}(:,:,i) = tmp_xi;
          end
          
        end
      end
      %for i = 1:size(t_gamma,2)
      %  gamma(:,i) = t_gamma(:,i);
      %end
      %gamma(:,1:tT) = t_gamma;      
      %sumxi = t_sumxi;
      
    end
    gamma_all(:,n,1:tT) = t_gamma(:,1:tT);
    xi_sum(:,:,n) = t_sumxi;
    
    % from scaling constants
    %phi_norm(n) = sum(log(t_c));

    % from scaling constants (more stable)
    phi_norm(n) = sum(log(t_c)) + sum(max_fb_logrho);

  end
end

% output
fbstats.logrho_Saved = logrho_Saved;
fbstats.gamma_all    = gamma_all;
fbstats.xi_sum       = xi_sum;
fbstats.phi_norm     = phi_norm;  %Phi (normalization constant)
fbstats.logLambdaTilde = logLambdaTilde;
fbstats.logPiTilde     = logPiTilde;
fbstats.logATilde      = logATilde;

if savexi
  fbstats.xi_Saved = xi_Saved;
end
